### ISSUE TITLE
ViewModel with inheritance in ASP .NET Core

### ISSUE THREAD

**Issue by raphaelm22** [aspnet/Mvc#6189]
_Note: Criei um repositório para esse questionamento: [https://github.com/raphaelm22/InheritanceViewModel](https://github.com/raphaelm22/InheritanceViewModel)_

Background
-------

In payment form, there are two methods. First is using the credit card, and the other method is using generics points, in this case I called "Gold".

The initial problem
--------
I do not want create one ViewModel with all fields to reference the Credit Card and data to access the "account of Gold". Because the user only need fill the fields for Credit Card or "account of Gold". The problem is more evident when is need create custom validation. 

Solution
--------
I belive the inheritance solve that. So I propose that the ViewModel be something like this:

```C#
public abstract class PaymentViewModelBase
{
    public abstract EPaymentType PaymentType { get; }
}

public class PaymentByCreditCardViewModel : PaymentViewModelBase
{
    public override EPaymentType PaymentType => EPaymentType.CreditCard;

    [Display(Name = "Name of credit card owner")]
    [Required()]
    [StringLength(100)]
    [RegularExpression("(?i)^[a-z ]+$")]
    public string CardOwnerName { get; set; }

    [Display(Name = "Number of credit card")]
    [Required()]
    [StringLength(16)]
    [RegularExpression("(?i)[0-9]{16}$")]
    public string CardNumber { get; set; }
}

public class PaymentByGoldViewModel : PaymentViewModelBase
{
    public override EPaymentType PaymentType => EPaymentType.Gold;
        
    [Display(Name = "Account ID")]
    [Required()]
    public string AccountId { get; set; }

    [Display(Name = "Account Token")]
    [Required()]
    public string AccountToken { get; set; }
}

```

The new problem
--------
Now, I have a new problem. There aren't a bind to object with inheritance. 

Solution
--------
So, I wrote a bind to resolve that. But it's a little dumb, you must manually say what the final type will be. Let's go to see.

First you decorate the abstract class with the ``InheritanceBinder`` and tells the class which will help you choose the final type of ViewModel.

```C#
[InheritanceBinder(typeof(PaymentViewModelBaseResolve))]
public abstract class PaymentViewModelBase
```

And then imply the ``Resolve`` method of the ``IInheritanceBinderResolve`` interface. 

The ``bindingContext`` attribute will probably have everything that will help you choose the final ViewModel type

```C#
public class PaymentViewModelBaseResolve : IInheritanceBinderResolve
{
    public Type Resolve(ModelBindingContext bindingContext)
    {
        var key = $"{bindingContext.ModelName}.{nameof(PaymentViewModelBase.PaymentType)}";
        var typeOfPayment = bindingContext.ValueProvider.GetValue(key);

        if (typeOfPayment == ValueProviderResult.None)
            throw new Exception($"The form must contain the property {nameof(PaymentViewModelBase.PaymentType)} to do a bind");

        EPaymentType paymentType;
        if (!Enum.TryParse(typeOfPayment.FirstValue, out paymentType))
            throw new Exception($"The property {nameof(PaymentViewModelBase.PaymentType)} don't have the supported value");

        switch (paymentType)
        {
            case EPaymentType.CreditCard:
                return typeof(PaymentByCreditCardViewModel);
            case EPaymentType.Gold:
                return typeof(PaymentByGoldViewModel);
            default:
                throw new Exception($"The type {paymentType} not been mapped");
        }
    }
}
```
In that case, the ViewModel has the PaymentType property, it is an enum, so depending on the value I know which end type.

The good news is when ``IsValid`` is invoked, it only validates the required properties, ie if the user informs the credit card it will only validate the ``PaymentByCreditCardViewModel`` fields.

**What do you think of this approach? In the repository has the bind implementation and a sample project.**

**Comment by frankabbruzzese** [comment]
Lightweight solution. However, it has the disadvantage the user must write an interface implementation foreach class/subclasses group.

[Here ](http://examples.aspnetcore.mvc-controls.com/InputExamples/SubClass) you find a solution where a `subclass ` tag-helper surround the fields that are specific to each subclass.  That fields are rendered either if the model is null or if a specific class of that is subclass is passed. 

When the model is null (no subclass choice has been already done) all subclass specific fields are rendered, and they bi hidden/shown with a tab contro where user select subclass it want to create (ie payment type). 

No need for extra-code. You only need to decorate the "acceptable" subclasses with the [RunTimeType] attribute. This is necessary to prevent denial of service attacks (without this server-side declaration the user might force the creation of instances of dangerous types).

The solution above works also if one submits data in json format, since it configures properly also the json.net deserializer.

**Comment by raphaelm22** [comment]
The solution as a whole is very interesting and I really liked the `taghelper `subclass.

But the disadvantage I encountered was when the user did not fill in the `RegisterNumber `or `Matr `property, then bind will build a `Person `object and it may be that the code is expecting a subclass, so in action there should be a treatment for that.

What I'm trying to say is that the two solutions have a similar problem and are they are treated in different places, which in the case of the proposed solution, this problem is handled in the `IInheritanceBinderResolve `interface.

Do you agree?

**Comment by frankabbruzzese** [comment]
@raphaelm22 ,
No, in my solution also if  `RegisterNumber` or `Matr` are submitted empty the right subclass is created by the custom model binder, that performs also the right validation. 

The job of your `InheritanceBindResolver `is somehow inferred automatically. Actually what is important is not the name of the fields but a code that is inserted in the name of any field that is surrounded by the subclass taghelper (use browser inspect to see the names of fileds in my example). That code identifies the subclass specified in the taghelper. 

When used in insertion mode (ie all fields are empty because a null model is passed) the  content all all subclass taghelpers is shown, so the developer must add some javascript to remove all fields but the ones corresponding to the subclass selected by the user. The user click some "Tabs" to select the right subclass and each tab button must re-add to the dom the right fields for the selected class.

When used in edit mode, only the fields corresponding to the right subclass are rendered. All other subclass taghelpers doesnt render their content because they detected the type passed differ from their type.

The job of the `InheritanceBindResolver ` is inferred automatically thanks to a preprocessing of all Types decorated with the [RunTimeTypeAttribute], that assign to all all involved subclasses an unique code.

Both solution solve the same problem. My solution is completely automatic, but require more customization of the default Asp.net core Mvc behavior, while your solution is more Lightweight but requires manual job (implementing all needed `InheritanceBindResolver `). 

Your solution allow for "custom" subclass customization since user write himself the discrimination function, while my discrimination function is "standard".  Not sure, however, if this flexibility is usefull in practice (at least I was not able to imagine an example that need some "atypical discrimination").

The vantage of having everything "automatic" is that the subclass "feature" may be used automatically by other controls like for instance a [grid](http://examples.aspnetcore.mvc-controls.com/CGrids/CustomServer). In the example the grid recognize automatically a subclass and renders the extra Price/Year column. Subclasses are also edited properly. Grid has two add buttons one for the normal class and another one for the subclass.

**Comment by raphaelm22** [comment]
Hi @frankabbruzzese, 

I had not noticed the attributes name/id when I used browser inspect, now I understand how work the classes `RequestTransformationsRegister` and `TransformationModelBinder `.

```html
<input name="Test1.0_.RegisterNumber" id="Test1_0__RegisterNumber" />
<input name="Test2.2_.Matr" id="Test2_2__Matr" />
```
I agree with you, I don't know say if it's worth paying the cost of customization of `IInheritanceBinderResolve`, but I have doubts too about whether automation meets all case in an elegant way.

**Comment by Eilon** [comment]
This is indeed an interesting problem. I think if I was doing this for my own site I would just have a single top-level model (no inheritance), and it would have two sub-properties, one for Credit Card, and one for Gold. The top-level model would have an `enum` property that would be either `PaymentType.CreditCard` or `PaymentType.Gold`, and the validators I would set up for the model would be aware of the dependencies between the sub-models and this enum.

As an aside, this question is probably a better fit for a site such as https://stackoverflow.com/.
